src/ORM/Connect/DBQueryBuilder.php

```diff
@@ -57,9 +61,73 @@ public function buildSQL(SQLExpression $query, &$parameters)
                 "Not implemented: query generation for type " . get_class($query)
             );
         }
+
+        if (Director::isDev()) {
+            $sql = $this->buildTraceComment() . $sql;
+        }
+
         return $sql;
     }
 
+    /**
+     * Builds an SQL comment indicating where the query was executed from.
+     */
+    protected function buildTraceComment(): string
+    {
+        $comment = '/* ';
+
+        $ignoreBaseClasses = [
+            self::class,
+            DataList::class,
+            DataQuery::class,
+            SQLExpression::class,
+            DB::class,
+            Database::class,
+            DBConnector::class,
+            DBSchemaManager::class,
+            TransactionManager::class,
+        ];
+
+        $line = null;
+        $file = null;
+        $class = null;
+        $function = null;
+
+        // Don't include arguments in the trace (since we don't need them), and only go back 15 levels.
+        // Anything further than that and we've probably over-abstracted things.
+        $trace = debug_backtrace(DEBUG_BACKTRACE_PROVIDE_OBJECT|DEBUG_BACKTRACE_IGNORE_ARGS, 15);
```

Comment:
Need to make sure that there's a level of 'opt-in' to this running because 99.99% of the time people DON'T want this to execute during regular local dev as we're doing things that are not working out exactly where SQL came from. debug_backtrace is a relatively expensive operation

Reply:
It's not that expensive.... but I'm fine with making this opt-in. I'll add a config variable and remove the dev-only requirement.